src/ray/gcs/redis_module/ray_redis_module.cc

```diff
@@ -20,33 +23,51 @@
 extern RedisChainModule module;
 #endif
 
-#define CHECK_ERROR(STATUS, MESSAGE)                   \
-  if ((STATUS) == REDISMODULE_ERR) {                   \
-    return RedisModule_ReplyWithError(ctx, (MESSAGE)); \
+#define CHECK_OR_REPLY(CONDITION, MESSAGE)      \
+  if (!(CONDITION)) {                           \
+    RedisModule_ReplyWithError(ctx, (MESSAGE)); \
+    return REDISMODULE_ERR;                     \
+  }
+
+// This macro can be used at the top level of redis module.
+#define REPLY_RETURN_IF_NOT_OK(STATUS)                           \
+  {                                                              \
+    auto status = (STATUS);                                      \
+    if (!status.ok()) {                                          \
+      RedisModule_ReplyWithError(ctx, status.message().c_str()); \
```

Comment:
According to [redis code](https://github.com/antirez/redis/blob/f5d918b2bb1bae3c84dee635ac13526779e9294c/src/module.c#L1071-L1079), `RedisModule_ReplyWithError` is not an async function. It will use the message immediately. After `RedisModule_ReplyWithError` returns, it is safe for `status` to go out of scope.

Reply:
The signature of `status.message` is

```
std::string message() const { return ok() ? "" : state_->msg; }
```

so I think `message` returns a temporary copy which gets destroyed immediately.

Reply:
I don't think so. `state_` will be deleted in the destructor of `Status`, which is called when `status` goes out of scope after `RedisModule_ReplyWithError` finishes. Actually, the returned string is a stack variable, the stack will keep growing while calling `RedisModule_ReplyWithError`. After `RedisModule_ReplyWithError` is finished, the temporary string will be destructed. The following code snipped will show that the temporary Test instance will  have the same address of `b`. However, the destructor is called after the function finishes.
```C++
class Test {
 public:
  Test() : msg("Test message.") {
    std::cout << "Test Constructor called:" << this << std::endl;
  }
  ~Test() {
    std::cout << "Test Destructor called." << std::endl;
  }
  std::string get() { return msg.c_str(); };
 private:
  std::string msg;
};

void Function(const char* msg) {
  std::cout << "Function called: " << msg << std::endl;
}

int main(int argc, char **argv) {
  int a = 0;
  std::cout << "Before calling Function:" << &a << std::endl;
  Function(Test().get().c_str());
  int b = 0;
  std::cout << "After calling Function:" << &b << std::endl;
  return 0;
}
```

Reply:
Ok, you may be right. However, I've definitely had issues with this kind of thing in the past. @mehrdadn any chance you could comment?

Reply:
Destructors only get called at the end of a statement, not after a subexpression is evalutaed. So this is fine as long as `RedisModule_ReplyWithError` does not attempt to store the pointer and use it later. From what ]I'm reading](/antirez/redis/issues/3625#issuecomment-261937314), the string is indeed copied internally before the function returns and the original isn't used afterward, so it's fine.

PS. note that variables are said to go out of scope, whereas values (like `status.message()`) are said to get destroyed.

Reply:
Ok, thanks @guoyuhong @mehrdadn!

Reply:
Yup! Also, note that the test case above won't work, because even if the destructor was called before `Function`, the raw memory contents could still happen to be unchanged, so you wouldn't know which is the case just because the string is printed correctly.

Instead, you'd want to check when the destructor is called; here's how:

	#include <string>
	#include <iostream>

	struct Str : std::string
	{
		Str() : std::string("hello")
		{ std::cout << "Str()" << std::endl; }
		~Str() { std::cout << "~Str()" << std::endl; }
	};
	int print(char const *s) { std::cout << s << std::endl; return 0; }
	int main() { print(Str().c_str()) + print(Str().c_str()); }

The output shows that the destructors are called at the end of the statement.

	Str()
	hello
	Str()
	hello
	~Str()
	~Str()

Reply:
@mehrdadn Thanks, this example is right one~